File: HockeyPlayoffs/Handlers/DateTimeHandler.m
//
//  DateTimeHandler.m
//  Hockey Playoffs
//
//  Created by Pierre-Marc Airoldi on 2014-04-07.
//  Copyright (c) 2015 Pierre-Marc Airoldi. All rights reserved.
//

#import "DateTimeHandler.h"
#import "FormatterHandler.h"
#import "StringHandler.h"

@implementation DateTimeHandler

#pragma Public Methods

+(NSString *)getDateForDate:(NSString *)date andTime:(NSString *)time {
    
    if (![StringHandler compareString:date withRegex:@"\\d{4}-\\d{2}-\\d{2}"]) {
        return @"";
    }
    
    NSDate *formattedDate = [[self class] convertToTimeZoneForDate:date andTime:time];
    
    NSDateFormatter *formatter = [FormatterHandler longDateFormatter];
    
    return [formatter stringFromDate:formattedDate];
}

+(NSString *)getTimeForDate:(NSString *)date andTime:(NSString *)time {
    
    if (![StringHandler compareString:time withRegex:@"\\d{2}:\\d{2}:\\d{2}"]) {
        return NSLocalizedString(@"time.tbd", nil);
    }
    
    NSDate *formattedDate = [[self class] convertToTimeZoneForDate:date andTime:time];
    
    NSDateFormatter *formatter = [FormatterHandler timeFormatter];
    
    return [formatter stringFromDate:formattedDate];
}

+(NSString *)getStringForDate:(NSDate *)date {
    
    if (date == nil) {
        return @"";
    }
    
    NSDateFormatter *formatter = [FormatterHandler fullDateFormatter];
    
    NSDate *convertedData = [[self class] convertFromTimeZoneForDate:date];
    
    return [formatter stringFromDate:convertedData];
}

#pragma Private Methods

+(NSDate *)convertToTimeZoneForDate:(NSString *)date andTime:(NSString *)time {

    NSDateFormatter *formatter = [FormatterHandler fullDateTimeFormatter];

    if (date.length == 0) {
        date = @"1970-01-01";
    }
    
    if (time.length == 0) {
        time = @"00:00:00";
    }
    
    NSDate * sourceDate = [formatter dateFromString:[NSString stringWithFormat:@"%@ %@", date, time]];
    
	NSTimeZone *sourceTimeZone = [NSTimeZone timeZoneWithAbbreviation:@"EST"];
	NSTimeZone *destinationTimeZone = [NSTimeZone defaultTimeZone];
    
    NSInteger sourceGMTOffset = [sourceTimeZone secondsFromGMTForDate:sourceDate];
    NSInteger destinationGMTOffset = [destinationTimeZone secondsFromGMTForDate:sourceDate];
    NSTimeInterval interval = destinationGMTOffset - sourceGMTOffset;
    
    NSDate *destinationDate = [[NSDate alloc] initWithTimeInterval:interval sinceDate:sourceDate];
    
    return destinationDate;
}

+(NSDate *)convertFromTimeZoneForDate:(NSDate *)date {
    
	NSTimeZone *sourceTimeZone = [NSTimeZone defaultTimeZone];
	NSTimeZone *destinationTimeZone = [NSTimeZone timeZoneWithAbbreviation:@"EST"];
    
    NSInteger sourceGMTOffset = [sourceTimeZone secondsFromGMTForDate:date];
    NSInteger destinationGMTOffset = [destinationTimeZone secondsFromGMTForDate:date];
    NSTimeInterval interval = destinationGMTOffset - sourceGMTOffset;
    
    NSDate *destinationDate = [[NSDate alloc] initWithTimeInterval:interval sinceDate:date];
    
    return destinationDate;
}

@end

File: HockeyPlayoffs/Extensions/DateTimeHandler.swift
import Foundation

class DateTimeHandler {

    static func getDate(forDate date: String, andTime time: String) -> String {

        if !StringHandler.compare(date, withRegex: "\\d{4}-\\d{2}-\\d{2}") {
            return ""
        }

        let formattedDate = self.convertToTimeZone(forDate: date, andTime: time)
        let formatter =  FormatterHandler.longDateFormatter()

        return formatter.string(from: formattedDate)
    }

    static func getTime(forDate date: String, andTime time: String) -> String {

        if !StringHandler.compare(time, withRegex: "\\d{2}:\\d{2}:\\d{2}") {
            return NSLocalizedString("time.tbd", comment: "Time to be determined")
        }

        let formattedDate = self.convertToTimeZone(forDate: date, andTime: time)
        let formatter = FormatterHandler.timeFormatter()

        return formatter.string(from: formattedDate)
    }

    static func getString(forDate date: Date) -> String {

        let convertedDate = self.convertFromTimeZone(forDate: date)
        let formatter = FormatterHandler.fullDateFormatter()

        return formatter.string(from: convertedDate)
    }

    private static func convertToTimeZone(forDate date: String, andTime time: String) -> Date {

        let formatter = FormatterHandler.fullDateTimeFormatter()
        let combinedDate = "\(date.count == 0 ? "1970-01-01" : date) \(time.count == 0 ? "00:00:00" : time)"

        guard let sourceDate = formatter.date(from: combinedDate), let sourceTimeZone = NSTimeZone(abbreviation: "EST") else {
            return Date(timeIntervalSince1970: 0)
        }

        let destinationTimeZone = NSTimeZone.default

        let sourceGMTOffset = sourceTimeZone.secondsFromGMT(for: sourceDate)
        let destinationGMTOffset = destinationTimeZone.secondsFromGMT(for: sourceDate)
        let interval = TimeInterval(destinationGMTOffset - sourceGMTOffset)

        let destinationDate = Date(timeInterval: interval, since: sourceDate)

        return destinationDate
    }

    private static func convertFromTimeZone(forDate date: Date) -> Date {

        guard let destinationTimeZone = NSTimeZone(abbreviation: "EST") else {
            return Date(timeIntervalSince1970: 0)
        }

        let sourceTimeZone = NSTimeZone.default

        let sourceGMTOffset =  sourceTimeZone.secondsFromGMT(for: date)
        let destinationGMTOffset = destinationTimeZone.secondsFromGMT(for: date)
        let interval = TimeInterval(destinationGMTOffset - sourceGMTOffset)

        let destinationDate = Date(timeInterval: interval, since: date)

        return destinationDate
    }
}
